File: backend/internal/qrflow/payloads.go
package qrflow

import (
	"encoding/json"
	"errors"
	"fmt"
	"net/url"
	"regexp"
	"strconv"
	"strings"
)

var (
	ErrUnknownPayloadKind = errors.New("unknown payload kind")
	ErrInvalidPayload     = errors.New("invalid payload data")
)

var emailPattern = regexp.MustCompile(`^[^\s@]+@[^\s@]+\.[^\s@]+$`)

func EncodePayload(kind string, raw []byte) (string, error) {
	switch kind {
	case "url":
		return encodeUrl(raw)
	case "text":
		return encodeText(raw)
	case "email":
		return encodeEmail(raw)
	case "phone":
		return encodePhone(raw)
	case "sms":
		return encodeSms(raw)
	case "wifi":
		return encodeWifi(raw)
	case "vcard":
		return encodeVCard(raw)
	case "location":
		return encodeLocation(raw)
	case "crypto-address":
		return encodeCryptoAddress(raw)
	default:
		return "", fmt.Errorf("%w: %q", ErrUnknownPayloadKind, kind)
	}
}

func parsePayload[T any](raw []byte) (T, error) {
	var payload T
	if err := json.Unmarshal(raw, &payload); err != nil {
		return payload, fmt.Errorf("%w: %v", ErrInvalidPayload, err)
	}
	return payload, nil
}

func requiredString(value string, label string) (string, error) {
	trimmed := strings.TrimSpace(value)
	if trimmed == "" {
		return "", fmt.Errorf("%w: %s is required", ErrInvalidPayload, label)
	}
	return trimmed, nil
}

func validateEmail(value string, label string) (string, error) {
	email, err := requiredString(value, label)
	if err != nil {
		return "", err
	}
	if !emailPattern.MatchString(email) {
		return "", fmt.Errorf("%w: %s must be a valid email address", ErrInvalidPayload, label)
	}
	return email, nil
}

type UrlPayload struct {
	DestinationUrl string `json:"destinationUrl"`
}

func encodeUrl(raw []byte) (string, error) {
	p, err := parsePayload[UrlPayload](raw)
	if err != nil {
		return "", err
	}
	parsed, err := ValidateDestination(p.DestinationUrl)
	if err != nil {
		return "", err
	}
	return parsed.String(), nil
}

type TextPayload struct {
	Text string `json:"text"`
}

func encodeText(raw []byte) (string, error) {
	p, err := parsePayload[TextPayload](raw)
	if err != nil {
		return "", err
	}
	text, err := requiredString(p.Text, "text")
	if err != nil {
		return "", err
	}
	return text, nil
}

type EmailPayload struct {
	To      string `json:"to"`
	Subject string `json:"subject,omitempty"`
	Body    string `json:"body,omitempty"`
}

func encodeEmail(raw []byte) (string, error) {
	p, err := parsePayload[EmailPayload](raw)
	if err != nil {
		return "", err
	}
	to, err := validateEmail(p.To, "to")
	if err != nil {
		return "", err
	}

	params := url.Values{}
	if p.Subject != "" {
		params.Set("subject", p.Subject)
	}
	if p.Body != "" {
		params.Set("body", p.Body)
	}

	query := params.Encode()
	if query != "" {
		return fmt.Sprintf("mailto:%s?%s", to, query), nil
	}
	return fmt.Sprintf("mailto:%s", to), nil
}

type PhonePayload struct {
	Number string `json:"number"`
}

func encodePhone(raw []byte) (string, error) {
	p, err := parsePayload[PhonePayload](raw)
	if err != nil {
		return "", err
	}
	number, err := requiredString(p.Number, "number")
	if err != nil {
		return "", err
	}
	return fmt.Sprintf("tel:%s", number), nil
}

type SmsPayload struct {
	Number  string `json:"number"`
	Message string `json:"message,omitempty"`
}

func encodeSms(raw []byte) (string, error) {
	p, err := parsePayload[SmsPayload](raw)
	if err != nil {
		return "", err
	}
	number, err := requiredString(p.Number, "number")
	if err != nil {
		return "", err
	}

	params := url.Values{}
	if p.Message != "" {
		params.Set("body", p.Message)
	}

	query := params.Encode()
	if query != "" {
		return fmt.Sprintf("sms:%s?%s", number, query), nil
	}
	return fmt.Sprintf("sms:%s", number), nil
}

type WifiPayload struct {
	SSID     string `json:"ssid"`
	Security string `json:"security,omitempty"`
	Password string `json:"password,omitempty"`
	Hidden   bool   `json:"hidden,omitempty"`
}

func escapeWifi(s string) string {
	s = strings.ReplaceAll(s, "\\", "\\\\")
	s = strings.ReplaceAll(s, ";", "\\;")
	s = strings.ReplaceAll(s, ",", "\\,")
	s = strings.ReplaceAll(s, ":", "\\:")
	s = strings.ReplaceAll(s, "\"", "\\\"")
	return s
}

func encodeWifi(raw []byte) (string, error) {
	p, err := parsePayload[WifiPayload](raw)
	if err != nil {
		return "", err
	}
	ssid, err := requiredString(p.SSID, "ssid")
	if err != nil {
		return "", err
	}

	auth := "WPA2"
	if p.Security != "" {
		switch p.Security {
		case "none":
			auth = "nopass"
		case "wpa", "wpa2", "wpa3":
			auth = strings.ToUpper(p.Security)
		default:
			return "", fmt.Errorf("%w: unsupported wifi security", ErrInvalidPayload)
		}
	}
	if auth != "nopass" && strings.TrimSpace(p.Password) == "" {
		return "", fmt.Errorf("%w: wifi password is required unless security is none", ErrInvalidPayload)
	}

	hidden := "false"
	if p.Hidden {
		hidden = "true"
	}

	return fmt.Sprintf("WIFI:T:%s;S:%s;P:%s;H:%s;;", auth, escapeWifi(ssid), escapeWifi(p.Password), hidden), nil
}

type VCardPayload struct {
	FullName     string `json:"fullName"`
	Organization string `json:"organization,omitempty"`
	Title        string `json:"title,omitempty"`
	Phone        string `json:"phone,omitempty"`
	Email        string `json:"email,omitempty"`
	URL          string `json:"url,omitempty"`
}

func escapeVCard(s string) string {
	s = strings.ReplaceAll(s, "\\", "\\\\")
	s = strings.ReplaceAll(s, "\n", "\\n")
	s = strings.ReplaceAll(s, ",", "\\,")
	s = strings.ReplaceAll(s, ";", "\\;")
	return s
}

func encodeVCard(raw []byte) (string, error) {
	p, err := parsePayload[VCardPayload](raw)
	if err != nil {
		return "", err
	}
	fullName, err := requiredString(p.FullName, "full name")
	if err != nil {
		return "", err
	}
	if p.Email != "" {
		if _, err := validateEmail(p.Email, "email"); err != nil {
			return "", err
		}
	}
	if p.URL != "" {
		if _, err := ValidateDestination(p.URL); err != nil {
			return "", err
		}
	}

	var lines []string
	lines = append(lines, "BEGIN:VCARD")
	lines = append(lines, "VERSION:3.0")
	lines = append(lines, fmt.Sprintf("FN:%s", escapeVCard(fullName)))

	if p.Organization != "" {
		lines = append(lines, fmt.Sprintf("ORG:%s", escapeVCard(p.Organization)))
	}
	if p.Title != "" {
		lines = append(lines, fmt.Sprintf("TITLE:%s", escapeVCard(p.Title)))
	}
	if p.Phone != "" {
		lines = append(lines, fmt.Sprintf("TEL:%s", escapeVCard(p.Phone)))
	}
	if p.Email != "" {
		lines = append(lines, fmt.Sprintf("EMAIL:%s", p.Email))
	}
	if p.URL != "" {
		lines = append(lines, fmt.Sprintf("URL:%s", p.URL))
	}

	lines = append(lines, "END:VCARD")
	return strings.Join(lines, "\n"), nil
}

type LocationPayload struct {
	Latitude  float64 `json:"latitude"`
	Longitude float64 `json:"longitude"`
	Label     string  `json:"label,omitempty"`
}

func encodeLocation(raw []byte) (string, error) {
	p, err := parsePayload[LocationPayload](raw)
	if err != nil {
		return "", err
	}
	if p.Latitude < -90 || p.Latitude > 90 || p.Longitude < -180 || p.Longitude > 180 {
		return "", fmt.Errorf("%w: invalid coordinates", ErrInvalidPayload)
	}

	lat := strconv.FormatFloat(p.Latitude, 'f', -1, 64)
	lon := strconv.FormatFloat(p.Longitude, 'f', -1, 64)
	base := fmt.Sprintf("geo:%s,%s", lat, lon)

	if p.Label != "" {
		return fmt.Sprintf("%s?q=%s,%s(%s)", base, lat, lon, url.QueryEscape(p.Label)), nil
	}
	return base, nil
}

type CryptoAddressPayload struct {
	Currency string `json:"currency"`
	Address  string `json:"address"`
	Amount   string `json:"amount,omitempty"`
	Label    string `json:"label,omitempty"`
}

func encodeCryptoAddress(raw []byte) (string, error) {
	p, err := parsePayload[CryptoAddressPayload](raw)
	if err != nil {
		return "", err
	}
	currency, err := requiredString(p.Currency, "currency")
	if err != nil {
		return "", err
	}
	address, err := requiredString(p.Address, "address")
	if err != nil {
		return "", err
	}
	if !isSupportedCryptoCurrency(currency) {
		return "", fmt.Errorf("%w: unsupported crypto currency", ErrInvalidPayload)
	}

	if currency == "btc" {
		params := url.Values{}
		if p.Amount != "" {
			params.Set("amount", p.Amount)
		}
		if p.Label != "" {
			params.Set("label", p.Label)
		}
		query := params.Encode()
		if query != "" {
			return fmt.Sprintf("bitcoin:%s?%s", address, query), nil
		}
		return fmt.Sprintf("bitcoin:%s", address), nil
	}

	return address, nil
}

func isSupportedCryptoCurrency(currency string) bool {
	switch currency {
	case "btc", "eth", "usdc", "sol", "ltc", "other":
		return true
	default:
		return false
	}
}

File: packages/qr-core/src/payloads.ts
export type QrPayloadKind =
  | "url"
  | "text"
  | "email"
  | "phone"
  | "sms"
  | "wifi"
  | "vcard"
  | "location"
  | "crypto-address";

export type UrlPayload = {
  destinationUrl: string;
};

export type TextPayload = {
  text: string;
};

export type EmailPayload = {
  to: string;
  subject?: string;
  body?: string;
};

export type PhonePayload = {
  number: string;
};

export type SmsPayload = {
  number: string;
  message?: string;
};

export type WifiPayload = {
  ssid: string;
  security?: "none" | "wpa" | "wpa2" | "wpa3";
  password?: string;
  hidden?: boolean;
};

export type VCardPayload = {
  fullName: string;
  organization?: string;
  title?: string;
  phone?: string;
  email?: string;
  url?: string;
};

export type LocationPayload = {
  latitude: number;
  longitude: number;
  label?: string;
};

export type CryptoAddressPayload = {
  currency: "btc" | "eth" | "usdc" | "sol" | "ltc" | "other";
  address: string;
  amount?: string;
  label?: string;
};

export type QrPayloadByKind = {
  url: UrlPayload;
  text: TextPayload;
  email: EmailPayload;
  phone: PhonePayload;
  sms: SmsPayload;
  wifi: WifiPayload;
  vcard: VCardPayload;
  location: LocationPayload;
  "crypto-address": CryptoAddressPayload;
};

export type QrPayloadConfigV1<K extends QrPayloadKind = QrPayloadKind> = {
  schemaVersion: "qurl.qr-payload-config.v1";
  kind: K;
  payload: QrPayloadByKind[K];
};

export type QrTypeDefinitionV1 = {
  kind: QrPayloadKind;
  label: string;
  category: "direct" | "contact" | "communication" | "network" | "location" | "commerce";
  status: "mvp";
  directEncoding: true;
  payloadSchemaId: string;
};

export type QrTypeRegistryV1 = {
  schemaVersion: "qurl.qr-type-registry.v1";
  types: readonly QrTypeDefinitionV1[];
};

const PAYLOAD_SCHEMA_ID = "https://qurl.dev/schemas/qr-payload-config.v1.schema.json";
const WEB_PROTOCOLS = new Set(["http:", "https:"]);
const EMAIL_PATTERN = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;

export const QR_TYPE_REGISTRY_V1: QrTypeRegistryV1 = {
  schemaVersion: "qurl.qr-type-registry.v1",
  types: [
    registryEntry("url", "Website URL", "direct"),
    registryEntry("text", "Plain text", "direct"),
    registryEntry("email", "Email", "communication"),
    registryEntry("phone", "Phone", "communication"),
    registryEntry("sms", "SMS", "communication"),
    registryEntry("wifi", "Wi-Fi", "network"),
    registryEntry("vcard", "Contact card", "contact"),
    registryEntry("location", "Location", "location"),
    registryEntry("crypto-address", "Crypto address", "commerce"),
  ],
};

function registryEntry(
  kind: QrPayloadKind,
  label: string,
  category: QrTypeDefinitionV1["category"],
): QrTypeDefinitionV1 {
  return {
    kind,
    label,
    category,
    status: "mvp",
    directEncoding: true,
    payloadSchemaId: `${PAYLOAD_SCHEMA_ID}#/$defs/${kind}`,
  };
}

function requiredString(value: unknown, label: string): string {
  if (typeof value !== "string") {
    throw new Error(`${label} is required.`);
  }

  const trimmed = value.trim();
  if (trimmed.length === 0) {
    throw new Error(`${label} is required.`);
  }

  return trimmed;
}

function optionalString(value: unknown): string | undefined {
  if (typeof value !== "string") {
    return undefined;
  }

  const trimmed = value.trim();
  return trimmed.length > 0 ? trimmed : undefined;
}

function normalizeWebUrl(input: string, label: string): string {
  const value = requiredString(input, label);
  const parsed = new URL(value);

  if (!WEB_PROTOCOLS.has(parsed.protocol)) {
    throw new Error(`${label} must use http or https.`);
  }

  return value;
}

function validateEmail(input: string, label = "Email"): string {
  const email = requiredString(input, label);
  if (!EMAIL_PATTERN.test(email)) {
    throw new Error(`${label} must be a valid email address.`);
  }

  return email;
}

function encodeQuery(values: Record<string, string | undefined>): string {
  const params = new URLSearchParams();
  Object.entries(values).forEach(([key, value]) => {
    if (value) {
      params.set(key, value);
    }
  });

  return params.toString();
}

function escapeWifiValue(value: string): string {
  return value.replace(/([\\;,":])/g, "\\$1");
}

function escapeVCardValue(value: string): string {
  return value
    .replace(/\\/g, "\\\\")
    .replace(/\n/g, "\\n")
    .replace(/,/g, "\\,")
    .replace(/;/g, "\\;");
}

export function createQrPayload<K extends QrPayloadKind>(
  kind: K,
  payload: QrPayloadByKind[K],
): QrPayloadConfigV1<K> {
  return {
    schemaVersion: "qurl.qr-payload-config.v1",
    kind,
    payload: normalizePayload(kind, payload),
  };
}

export function normalizePayload<K extends QrPayloadKind>(
  kind: K,
  payload: QrPayloadByKind[K],
): QrPayloadByKind[K] {
  switch (kind) {
    case "url":
      return {
        destinationUrl: normalizeWebUrl((payload as UrlPayload).destinationUrl, "Direct URL"),
      } as QrPayloadByKind[K];
    case "text":
      return { text: requiredString((payload as TextPayload).text, "Text") } as QrPayloadByKind[K];
    case "email":
      return {
        to: validateEmail((payload as EmailPayload).to),
        subject: optionalString((payload as EmailPayload).subject),
        body: optionalString((payload as EmailPayload).body),
      } as QrPayloadByKind[K];
    case "phone":
      return {
        number: requiredString((payload as PhonePayload).number, "Phone number"),
      } as QrPayloadByKind[K];
    case "sms":
      return {
        number: requiredString((payload as SmsPayload).number, "SMS number"),
        message: optionalString((payload as SmsPayload).message),
      } as QrPayloadByKind[K];
    case "wifi": {
      const wifi = payload as WifiPayload;
      const security = wifi.security ?? "wpa2";
      if (security !== "none" && optionalString(wifi.password) === undefined) {
        throw new Error("Wi-Fi password is required unless security is none.");
      }

      return {
        ssid: requiredString(wifi.ssid, "Wi-Fi network name"),
        security,
        password: security === "none" ? undefined : optionalString(wifi.password),
        hidden: Boolean(wifi.hidden),
      } as QrPayloadByKind[K];
    }
    case "vcard": {
      const card = payload as VCardPayload;
      return {
        fullName: requiredString(card.fullName, "Contact name"),
        organization: optionalString(card.organization),
        title: optionalString(card.title),
        phone: optionalString(card.phone),
        email: card.email ? validateEmail(card.email) : undefined,
        url: card.url ? normalizeWebUrl(card.url, "Contact URL") : undefined,
      } as QrPayloadByKind[K];
    }
    case "location": {
      const { latitude, longitude, label } = payload as LocationPayload;
      if (!Number.isFinite(latitude) || latitude < -90 || latitude > 90) {
        throw new Error("Latitude must be between -90 and 90.");
      }
      if (!Number.isFinite(longitude) || longitude < -180 || longitude > 180) {
        throw new Error("Longitude must be between -180 and 180.");
      }

      return { latitude, longitude, label: optionalString(label) } as QrPayloadByKind[K];
    }
    case "crypto-address": {
      const crypto = payload as CryptoAddressPayload;
      if (!["btc", "eth", "usdc", "sol", "ltc", "other"].includes(crypto.currency)) {
        throw new Error("Crypto currency is not supported.");
      }

      return {
        currency: crypto.currency,
        address: requiredString(crypto.address, "Crypto address"),
        amount: optionalString(crypto.amount),
        label: optionalString(crypto.label),
      } as QrPayloadByKind[K];
    }
  }

  throw new Error(`Unsupported QR payload kind: ${kind satisfies never}`);
}

export function encodeQrPayload(config: QrPayloadConfigV1): string {
  const payload = normalizePayload(config.kind, config.payload);

  switch (config.kind) {
    case "url":
      return (payload as UrlPayload).destinationUrl;
    case "text":
      return (payload as TextPayload).text;
    case "email": {
      const email = payload as EmailPayload;
      const query = encodeQuery({ subject: email.subject, body: email.body });
      return `mailto:${email.to}${query ? `?${query}` : ""}`;
    }
    case "phone":
      return `tel:${(payload as PhonePayload).number}`;
    case "sms": {
      const sms = payload as SmsPayload;
      const query = encodeQuery({ body: sms.message });
      return `sms:${sms.number}${query ? `?${query}` : ""}`;
    }
    case "wifi": {
      const wifi = payload as WifiPayload;
      const auth = wifi.security === "none" ? "nopass" : (wifi.security ?? "wpa2").toUpperCase();
      return `WIFI:T:${auth};S:${escapeWifiValue(wifi.ssid)};P:${escapeWifiValue(wifi.password ?? "")};H:${wifi.hidden ? "true" : "false"};;`;
    }
    case "vcard": {
      const card = payload as VCardPayload;
      const lines = ["BEGIN:VCARD", "VERSION:3.0", `FN:${escapeVCardValue(card.fullName)}`];
      if (card.organization) lines.push(`ORG:${escapeVCardValue(card.organization)}`);
      if (card.title) lines.push(`TITLE:${escapeVCardValue(card.title)}`);
      if (card.phone) lines.push(`TEL:${escapeVCardValue(card.phone)}`);
      if (card.email) lines.push(`EMAIL:${card.email}`);
      if (card.url) lines.push(`URL:${card.url}`);
      lines.push("END:VCARD");
      return lines.join("\n");
    }
    case "location": {
      const loc = payload as LocationPayload;
      const base = `geo:${loc.latitude},${loc.longitude}`;
      return loc.label
        ? `${base}?q=${loc.latitude},${loc.longitude}(${encodeURIComponent(loc.label)})`
        : base;
    }
    case "crypto-address": {
      const crypto = payload as CryptoAddressPayload;
      if (crypto.currency === "btc") {
        const query = encodeQuery({ amount: crypto.amount, label: crypto.label });
        return `bitcoin:${crypto.address}${query ? `?${query}` : ""}`;
      }

      return crypto.address;
    }
  }

  throw new Error(`Unsupported QR payload kind: ${config.kind satisfies never}`);
}
